write.go

```diff
@@ -38,3 +85,391 @@ func toOptSet(opts ...WriteOption) *writeOptSet {
 	}
 	return optSet
 }
+
+func writeFileHeader(w dicomio.Writer, ds *Dataset, metaElems []*Element, opts ...WriteOption) error {
+	w.SetTransferSynax(binary.LittleEndian, false)
+
+	subWriter := dicomio.NewWriter(&bytes.Buffer{}, binary.LittleEndian, false)
+	tagsUsed := make(map[tag.Tag]bool)
+	tagsUsed[tag.FileMetaInformationGroupLength] = true
+
+	// TODO make better structure for error checking so it's no so many lines
+	err := writeMetaElem(subWriter, tag.MediaStorageSOPClassUID, ds, &tagsUsed, opts...)
+	if err != nil {
+		return err
+	}
+	err = writeMetaElem(subWriter, tag.MediaStorageSOPInstanceUID, ds, &tagsUsed, opts...)
+	if err != nil {
+		return err
+	}
+	err = writeMetaElem(subWriter, tag.TransferSyntaxUID, ds, &tagsUsed, opts...)
+	if err != nil {
+		return err
+	}
+
+	for _, elem := range metaElems {
+		if elem.Tag.Group == tag.MetadataGroup {
+			if _, ok := tagsUsed[elem.Tag]; !ok {
+				err = writeElement(subWriter, elem, opts...)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	metaBytes := subWriter.Bytes()
+	w.WriteZeros(128)
+	w.WriteString("DICM")
+	lengthElem, err := newElement(tag.FileMetaInformationGroupLength, []int{len(metaBytes)})
+	if err != nil {
+		return err
+	}
+
+	err = writeElement(w, lengthElem, opts...)
+	if err != nil {
+		return err
+	}
+	w.WriteBytes(metaBytes)
+
+	return nil
+}
+
+func writeElement(w dicomio.Writer, elem *Element, opts ...WriteOption) error {
+	// parse WriteOption options
+	options := toOptSet(opts...)
+	vr := elem.RawValueRepresentation
+	var err error // to fix 'declared and not used' errors
```

Comment:
There's a weird bug in Go that I came across where it will treat the variables inside of nests slightly weirdly with assignment. The explicit declaration of `err` in this case is actually necessary to prevent `./write.go:147:3: vr declared but not used`. This happens (I think) because Go sees an assignment operator (`:=`) inside of the first if statement, which is a different scope than the originally declared `vr` in the base function scope, and then assigns a new `vr` with a different memory address inside the if statement, which of course isn't used, and the compiler spits out a "not used" error.

```
 vr := elem.RawValueRepresentation
        // var err error // to fix 'declared and not used' errors
        // SkipVRVerification
        if !options.skipVRVerification {
                vr, err := verifyVR(elem.Tag, elem.RawValueRepresentation)    // vr is redeclared and not used in this scope
                if err != nil {
                        return err
                }
               // error occurs here because reached end of scope and redeclared vr not used
        }
```

Having the explicit declaration of `err` makes there no need to use the assignment operator (`:=`) and thus avoid the problem of accidentally redeclaring `vr`.